updating CONTRIBUTING.md (#11920)
Azure/azure-sdk-for-js: CONTRIBUTING.md

## Project orchestration

This project uses [Rush](https://rushjs.io) to manage many of our Azure SDK libraries within a single repository. It is highly recommended that you read the [Rush Developer Tutorials](https://rushjs.io/pages/developer/new_developer/) to familiarize yourself with the tool.

Rush provides many benefits:

> - Some of our devDependencies are not published to the public registery (.e.g. our own ESLint plugin), and Rush is configured to install them correctly.

Reviewer: clean some small typos 
```suggestion
- Some of our devDependencies are not published to the public registry (e.g. our ESLint plugin), and Rush is configured to install them correctly.
```

- Rush uses [PNPM](https://pnpm.js.org) to install all dependencies across the SDK. Together they solve problems involving [phantom dependencies](https://rushjs.io/pages/advanced/phantom_deps/) and [NPM doppelgangers](https://rushjs.io/pages/advanced/npm_doppelgangers/). The way PNPM lays out packages also ensures that you can never accidentally use a dependency you don't directly declare in your package.json.
- Dependencies between different libraries within the Azure SDK will be locally linked by default. This means you can make a local change in a library your library depends on, and it will just work without needing to use awkward "file:" paths in your package.json.
- When a change is made in a local dependency, Rush will detect that the dependency is dirty and will rebuild it if you attempt to build a project that consumes that dependency.
- Rush runs project tasks in parallel, subject to the inter-project dependencies that it detects. It also performs incremental builds by default, not rebuilding anything unnecessary (unless you tell it to).

> Not every library in the repository is managed by Rush yet. Those packages can still be managed using `npm`.

Reviewer 2: Can we add a note on how does one identify whether a package is managed by Rush or not?

Author: done.

> Not every library in the repository is managed by Rush yet, only those listed in the `projects` property in [rush.json](../rush.json). Packages not managed by Rush can still be managed using `npm`.

Reviewer 3: This change is causing link verification failure. Can you please replace this relative URL to absolute URL. Check here for more information: https://aka.ms/azsdk/guideline/links

This project uses [Rush](https://rushjs.io) to manage our many Azure SDK libraries within a single repository. It is highly recommended that you read the [Rush Developer Tutorials](https://rushjs.io/pages/developer/new_developer/) to familiarize yourself with the tool.

> While you can using the standard `npm` workflow, adopting Rush will provide you many benefits:

Reviewer 2: ```suggestion
While you can still use the standard `npm` workflow, adopting Rush will provide you many benefits:
```

> - Some of our devDependencies are not published to the public registery (.e.g. our own ESLint plugin), and Rush is configured to install them correctly.

Reviewer 2: How would someone using npm and not rush handle this case?

Author: I do not think there is a way to handle this case with `npm` unless we publish our internal tools somewhere. We need to either maintain our own registery or publish them [privately](https://docs.npmjs.com/creating-and-publishing-private-packages).

Reviewer 2: In that case should we avoid talking about the option of not using rush and using only npm? Or perhaps that npm is applicable to the auto generated packages and rush for the rest?

Author: I did an edit here: https://github.com/Azure/azure-sdk-for-js/pull/11920/commits/931130decf89f699c5aacfb1f6ff53fa40c21dde.

## Setting up your environment

> Want to get started hacking on the code? Super!

Reviewer 2: This sounds like it has an abrupt stop, can we have the below?

```suggestion
Want to get started hacking on the code? Great! Keep reading.
```

### Using Visual Studio Code

We love [Visual Studio Code](https://code.visualstudio.com/) for many reasons, mainly:

- You can debug JavaScript/TypeScript code right away with [automatic debugging configuration](https://code.visualstudio.com/updates/v1_45#_automatic-debug-configurations).

> - You can use it with Github's [Codespaces](https://visualstudio.microsoft.com/services/github-codespaces/) to develop inside a docker container that has all the prerequisites.

Reviewer 2: ```suggestion
- You can use it with GitHub's [Codespaces](https://visualstudio.microsoft.com/services/github-codespaces/) to develop inside a docker container that has all the prerequisites.
```

- You get [excellent support for TypeScript](https://code.visualstudio.com/Docs/languages/typescript).

### Prerequisites

> If you do not use github's Codespaces, make sure you have the prerequisites installed and available on your `$PATH`:

Reviewer 2: > If you do not use...

This would be a valid statement to make if we have already spoken about how to use codespaces with this repo. Are there any instructions we can add for that? Then we can update the above to be

```suggestion
If you are not using GitHub's Codespaces, make sure you have the prerequisites installed and available on your `$PATH`:
```

Author: I pushed this change: https://github.com/Azure/azure-sdk-for-js/pull/11920/commits/e8de80e5cc847fa48ba0b5004b250883c821153e.

3. Open a terminal and move into your local copy (`cd azure-sdk-for-js`)
4. Install and link all dependencies (`rush update`)
5. Build the code base (`rush build`)

### Making the switch

If you have previously worked in this repo using the `npm` workflow, the first time you switch to using Rush you should commit or stash any untracked files and then get back to a clean state by running `rush reset-workspace` before proceeding any further. This will get rid of any latent package-lock files, as well as your existing (incompatible) node_modules directories. You can then proceed down the path outlined below.

### Using Visual Studio Code

#### Debugging

Debugging Node.js code in VSCode is [well documented](https://code.visualstudio.com/docs/nodejs/nodejs-debugging). However, starting from version 1.45.1, VSCode can automatically debug Node.js code in most cases without having to write custom `launch.json` files for that purpose and this is true for our SDKs code as well. A demonstration of that feature can be found in the [release notes](https://code.visualstudio.com/updates/v1_45#_automatic-debug-configurations).

#### Warnings

> Visual Studio Code has a feature which will automatically fetch and install @types packages for you, using the standard npm package manager. This will cause problems with your node_modules directory, since Rush uses PNPM which lays out this directory quite differently. It's highly recommended that you ensure "Typescript: Disable Automatic Type Acquisition" is checked in your VSCode Workspace Settings (or ensure `typescript.disableAutomaticTypeAcquisition` is present in your .vscode/settings.json file).

Author: this setting is already there in the `.vscode/settings.json`, so no need to tell people about it.

> Visual Studio Code has a feature which will automatically fetch and install @types packages for you, using the standard npm package manager. This will cause problems with your node_modules directory, since Rush uses PNPM which lays out this directory quite differently. It's highly recommended that you ensure "Typescript: Disable Automatic Type Acquisition" is checked in your VSCode Workspace Settings (or ensure `typescript.disableAutomaticTypeAcquisition` is present in your .vscode/settings.json file).

Author: this setting is already there in the .vscode/settings.json, so no need to tell people about it.

The current version of VSCode for Windows has a bug that may cause a "file locked" error when you run any Rush command that modifies your node_modules directory:

```
ERROR: Error: Error: EPERM: operation not permitted, mkdir 'C:\XXXXX\node_modules'
Often this is caused by a file lock from a process such as your text editor, command prompt, or "gulp serve"
```

> This bug is fixed in the Insiders build of VSCode (1.34), and will be included in the next release. Until then, you can resolve this by running the "Typescript: Restart TS server" command from the Command Palette to release the lock on the files.

Author: I guess this has been already fixed now.

### Warnings for Windows users

> Git for Windows has a bug where repository files may be unintentionally removed by `git clean -df` when a directory is locally linked. Because Rush creates local links between packages, you may encounter this. It's highly recommended to use the `rush reset-workspace` command to get your working directory back to a clean state instead. If you prefer to run `git clean -df` manually, you must first run `rush unlink` so that the operation can be performed safely.

Author: This is already discussed later in the doc.

## Inner loop developer workflow with Rush
## Development Workflows

### Installing and managing dependencies

> Run `rush update` to install the current set of package dependencies in all projects inside the repo.

Author: `rush update` is discussed in the next two paragraphs.

  - Rush will automatically manage the specific version needed by this repo as long as you have any v5 version installed.
  - If you're unable to install a global tool, you can instead call the wrapper script `node <repo root>/common/scripts/install-run-rush.js` any time the guide instructs you to run `rush`. The wrapper script will install a managed copy of Rush in a temporary directory for you.

### Building our repository

1. Fork this repo
2. Clone your fork locally (`git clone https://github.com/<youruser>/azure-sdk-for-js.git`)
3. Open a terminal and move into your local copy (`cd azure-sdk-for-js`)
4. Install and link all dependencies (`rush update`)

> 5. Build the code base (`rush build`)

Reviewer: should we make this `rush rebuild` just to future-proof it when folks re-run the command?

Author: hmm, what is wrong with incremental build the second time? both commands are described in the `Building` section.

Reviewer: when I joined the team I was told to never use `rush build` because sometimes the incremental build doesn't detect changes and you'll pull your hair out, but maybe that's dated advice?

Author: That is not my experience so far but I can believe incremental build might cause some weird state if it does not track everything. I will change it to `rush rebuild`.

> 5. Build the code base (`rush rebuild`)

Reviewer 2: We should fork the instructions at this point. Step 4 and 5 assume that the package is managed by rush

Author: done.

#### Warnings

Visual Studio Code has a feature which will automatically fetch and install @types packages for you, using the standard npm package manager. This will cause problems with your node_modules directory, since Rush uses PNPM which lays out this directory quite differently. It's highly recommended that you ensure "Typescript: Disable Automatic Type Acquisition" is checked in your VSCode Workspace Settings (or ensure `typescript.disableAutomaticTypeAcquisition` is present in your .vscode/settings.json file).

The current version of VSCode for Windows has a bug that may cause a "file locked" error when you run any Rush command that modifies your node_modules directory:

```
ERROR: Error: Error: EPERM: operation not permitted, mkdir 'C:\XXXXX\node_modules'
Often this is caused by a file lock from a process such as your text editor, command prompt, or "gulp serve"
```

This bug is fixed in the Insiders build of VSCode (1.34), and will be included in the next release. Until then, you can resolve this by running the "Typescript: Restart TS server" command from the Command Palette to release the lock on the files.

### Warnings for Windows users

> Git for Windows has a bug where repository files may be unintentionally removed by `git clean -df` when a directory is locally linked. Because Rush creates local links between packages, you may encounter this. It's highly recommended to use the `rush reset-workspace` command to get your working directory back to a clean state instead. If you prefer to run `git clean -df` manually, you must first run `rush unlink` so that the operation can be performed safely.

Reviewer: I don't see `rush reset-workspace` documented anymore - can we preserve that part somewhere? It seems like a useful command to fix problems.

Author: it is mentioned in the `Getting back to a clean state` section here.

Reviewer: thanks for pointing that out - somehow Edge's "find in page" missed that

## Onboarding a new library

All libraries must follow our [repository structure](https://github.com/Azure/azure-sdk/blob/master/docs/policies/repostructure.md) (specifically, it must be located at `sdk/<servicename>/<packagename>`) and your library's `package.json` must contain the required scripts as documented [above](#other-npm-scripts).

> The repository contains two different sets of libraries, each follows different rules for development and maintaining. The first type is created automatically from the swagger specifications and their code should not be edited. Onboarding such library is just a matter of pushing its auto-generated directory to the right location in the repository.

Reviewer 2: ```suggestion
The repository contains two different sets of libraries, each follows different rules for development and maintaining. The first type is generated automatically from the [swagger specifications](https://github.com/Azure/azure-rest-api-specs) and their code should not be edited by hand. Onboarding such library is just a matter of pushing its auto-generated directory to the right location in the repository.
```
